raysect/core/math/function/float/function1d/interpolate.pyx

```diff
@@ -0,0 +1,538 @@
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""
+Interpolation functions for float.Function1D
+
+Interpolators are accessed through interface class Interpolate1D, which
+"""
+
+import numpy as np
+cimport cython
+from raysect.core.math.cython.interpolation.linear cimport linear1d
+from raysect.core.math.cython.interpolation.cubic cimport calc_coefficients_1d, evaluate_cubic_1d
+from raysect.core.math.cython.utility cimport find_index, lerp
+
+DEF INT_LINEAR = 0
+DEF INT_CUBIC = 1
+DEF INT_CUBIC_CONSTRAINED = 2
+
+_INTERPOLATION_TYPES = {
+    'linear': INT_LINEAR,
+    'cubic': INT_CUBIC,
+    'cubic_constrained': INT_CUBIC_CONSTRAINED
+}
+
+DEF EXT_NONE = 0
+DEF EXT_NEAREST = 1
+DEF EXT_LINEAR = 2
+DEF EXT_QUADRATIC = 3
+
+_EXTRAPOLATION_TYPES = {
+    'none': EXT_NONE,
+    'nearest': EXT_NEAREST,
+    'linear': EXT_LINEAR,
+    'quadratic': EXT_QUADRATIC
+}
+
+cdef class Interpolate1D(Function1D):
+    """
+    Interface class for Function1D interpolators
+
+    :param object x: 1D array-like object of real values.
+    :param object f: 1D array-like object of real values.
+    :param str interpolation_type: Type of interpolation to use ('linear', 'cubic', 'cubic_constrained')
+    :param str extrapolation_type: Type of extrapolation to use ('none', 'nearest', 'linear', 'quadratic')
+    """
+    def __init__(self, object x, object f, str interpolation_type,
+                 str extrapolation_type, double extrapolation_range):
+
+        interpolation_type = interpolation_type.lower()
+        if  interpolation_type not in _INTERPOLATION_TYPES:
+            raise ValueError(f'Interpolation type {interpolation_type} not found.')
+
+        cdef int interpolator_idx = _INTERPOLATION_TYPES.get(interpolation_type)
+
+        extrapolation_type = extrapolation_type.lower()
+        if extrapolation_type not in _EXTRAPOLATION_TYPES:
+            raise ValueError(f'Extrapolation type {extrapolation_type} not found.')
+
+        cdef int extrapolator_idx = _EXTRAPOLATION_TYPES.get(extrapolation_type)
+
+        self.x = np.array(x, dtype=np.float64)
+        self.f = np.array(f, dtype=np.float64)
+        self._x = x
+        self._f = f
+        self._last_index = self._x.shape[0] -1
+        self._extrapolation_range = extrapolation_range
+
+        # dimensions checks
+        if x.ndim != 1:
+            raise ValueError(f'The x array must be 1D. Got {x.shape}.')
+
+        if f.ndim != 1:
+            raise ValueError(f'The x array must be 1D. Got {f.shape}.')
+
+        if x.shape != f.shape:
+            raise ValueError(f'Shape mismatch between x array ({x.shape}) and f array ({f.shape}).')
+
+        # test monotonicity
+        if (np.diff(x) <= 0).any():
+            raise ValueError('The x array must be monotonically increasing.')
+
+        # create extrapolator per extrapolation_type argument
+        if extrapolator_idx == EXT_NONE:
```

Comment:
Replace this with a dictionary lookup: { name: Class }

Reply:
Might be cleaner to add the name of each interpolator/extrapolator as a class attribute and then simply look them up e.g.:

```
cdef class MyExtrap:
    cdef readonly str ID = 'myextrap'

id_to_extrapolator = {
    MyExtrap.ID: MyExtrap
    ....
}

# get id string of any sub-object:
print(f'The strong id of this internal object is: {my_extrap.ID}')
```

Reply:
Agreed. However, Cython doesn't seem to support static member variables. 
I wonder if Python types would be feasible solution?

```cython
cdef class _Extrapolator1D:
    ID = NotImplemented

cdef class MyExtrap(_Extrapolator1D):
    ID = 'myextrap'

id_to_extrapolator = {
    MyExtrap.ID: MyExtrap
}
```

Alternatives that came to my mind:

1. Static methods

```cython
cdef class _Extrapolator1D:
 
   @staticmethod
     cdef str id():
           raise NotImplementedError('This is a base class..')

cdef class MyExtrap(_Extrapolator1D):
 
   @staticmethod
    cdef str id():
           return 'myextrap'

id_to_extrapolator = {
    MyExtrap.id(): MyExtrap
}
```

2. @classmethod + @cython.binding(True) instead of static